### PR TITLE
New resource `azurerm_api_management_tag`

### PR DESCRIPTION
**Pull request by yupwei68**
Fix https://github.com/terraform-providers/terraform-provider-azurerm/issues/10833

--- PASS: TestAccApiManagementTag_basic (274.12s)
--- PASS: TestAccApiManagementTag_requiresImport (289.36s)
--- PASS: TestAccApiManagementTag_update (416.31s)

### CODE REVIEW

File: azurerm/internal/services/apimanagement/api_management_tag_resource.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+	ctx, cancel := timeouts.ForCreateUpdate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	apiManagementId, err := parse.ApiManagementID(d.Get("api_management_id").(string))
+	if err != nil {
+		return err
+	}
+	name := d.Get("name").(string)
+
+	id := parse.NewTagID(subscriptionId, apiManagementId.ResourceGroup, apiManagementId.ServiceName, name)
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_tag_resource.go:75]:
```suggestion

	id := parse.NewTagID(subscriptionId, apiManagementId.ResourceGroup, apiManagementId.ServiceName, d.Get("name").(string))
```

File: azurerm/internal/services/apimanagement/api_management_tag_resource.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+	if d.IsNewResource() {
+		existing, err := client.Get(ctx, apiManagementId.ResourceGroup, apiManagementId.ServiceName, name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("checking for presence of existing Tag %q: %s", id, err)
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_tag_resource.go:81]:
```suggestion
				return fmt.Errorf("checking for presence of existing %s: %s", id, err)
```

File: azurerm/internal/services/apimanagement/api_management_tag_resource.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	}
+
+	parameters := apimanagement.TagCreateUpdateParameters{
+		TagContractProperties: &apimanagement.TagContractProperties{
+			DisplayName: &displayName,
+		},
+	}
+
+	if _, err := client.CreateOrUpdate(ctx, apiManagementId.ResourceGroup, apiManagementId.ServiceName, name, parameters, ""); err != nil {
+		return fmt.Errorf("creating/updating %q: %+v", id, err)
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_tag_resource.go:102]:
```suggestion
		return fmt.Errorf("creating/updating %s: %+v", id, err)
```

File: azurerm/internal/services/apimanagement/api_management_tag_resource.go (hunk 2 above)
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_tag_resource.go:81]:
```suggestion
				return fmt.Errorf("checking for presence of existing %s: %+v", id, err)
```

File: azurerm/internal/services/apimanagement/api_management_tag_resource.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	apiManagementId, err := parse.ApiManagementID(d.Get("api_management_id").(string))
+	if err != nil {
+		return err
+	}
+	name := d.Get("name").(string)
+
+	id := parse.NewTagID(subscriptionId, apiManagementId.ResourceGroup, apiManagementId.ServiceName, name)
+
+	if d.IsNewResource() {
+		existing, err := client.Get(ctx, apiManagementId.ResourceGroup, apiManagementId.ServiceName, name)
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_tag_resource.go:78]:
```suggestion
		existing, err := client.Get(ctx, apiManagementId.ResourceGroup, apiManagementId.ServiceName, id.Name)
```

File: azurerm/internal/services/apimanagement/api_management_tag_resource.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("checking for presence of existing Tag %q: %s", id, err)
+			}
+		}
+
+		if !utils.ResponseWasNotFound(existing.Response) {
+			return tf.ImportAsExistsError("azurerm_api_management_tag", id.ID())
+		}
+	}
+	displayName := d.Get("name").(string)
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_tag_resource.go:89]:
```suggestion
	displayName := name
```

File: azurerm/internal/services/apimanagement/api_management_tag_resource.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+		displayName = v.(string)
+	}
+
+	parameters := apimanagement.TagCreateUpdateParameters{
+		TagContractProperties: &apimanagement.TagContractProperties{
+			DisplayName: &displayName,
+		},
+	}
+
+	if _, err := client.CreateOrUpdate(ctx, apiManagementId.ResourceGroup, apiManagementId.ServiceName, name, parameters, ""); err != nil {
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_tag_resource.go:101]:
```suggestion
	if _, err := client.CreateOrUpdate(ctx, apiManagementId.ResourceGroup, apiManagementId.ServiceName, id.Name, parameters, ""); err != nil {
```

File: azurerm/internal/services/apimanagement/api_management_tag_resource.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+		return fmt.Errorf("creating/updating %q: %+v", id, err)
+	}
+
+	d.SetId(id.ID())
+
+	return resourceApiManagementTagRead(d, meta)
+}
+
+func resourceApiManagementTagRead(d *pluginsdk.ResourceData, meta interface{}) error {
+	subscriptionId := meta.(*clients.Client).Account.SubscriptionId
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_tag_resource.go:111]:
```suggestion
```

File: azurerm/internal/services/apimanagement/api_management_tag_resource.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[DEBUG] %q was not found - removing from state!", id)
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("retrieving %q: %+v", id, err)
+	}
+
+	d.Set("api_management_id", parse.NewApiManagementID(subscriptionId, id.ResourceGroup, id.ServiceName).ID())
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_tag_resource.go:132]:
```suggestion
	d.Set("api_management_id", parse.NewApiManagementID(id.SubscriptionId, id.ResourceGroup, id.ServiceName).ID())
```

File: azurerm/internal/services/apimanagement/api_management_tag_resource.go [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+	id, err := parse.TagID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Get(ctx, id.ResourceGroup, id.ServiceName, id.Name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[DEBUG] %q was not found - removing from state!", id)
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_tag_resource.go:124]:
```suggestion
			log.Printf("[DEBUG] %s was not found - removing from state!", id)
```

File: azurerm/internal/services/apimanagement/api_management_tag_resource.go [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("retrieving %q: %+v", id, err)
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_tag_resource.go:129]:
```suggestion
		return fmt.Errorf("retrieving %s: %+v", id, err)
```

File: azurerm/internal/services/apimanagement/api_management_tag_resource.go [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+	ctx, cancel := timeouts.ForDelete(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.TagID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	if _, err = client.Delete(ctx, id.ResourceGroup, id.ServiceName, id.Name, ""); err != nil {
+		return fmt.Errorf("deleting %q: %+v", id, err)
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_tag_resource.go:153]:
```suggestion
		return fmt.Errorf("deleting %s: %+v", id, err)
```

File: azurerm/internal/services/apimanagement/api_management_tag_resource_test.go [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+func (ApiManagementTagResource) Exists(ctx context.Context, clients *clients.Client, state *pluginsdk.InstanceState) (*bool, error) {
+	id, err := parse.TagID(state.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := clients.ApiManagement.TagClient.Get(ctx, id.ResourceGroup, id.ServiceName, id.Name)
+	if err != nil {
+		return nil, fmt.Errorf("reading %q: %+v", id, err)
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_tag_resource_test.go:86]:
```suggestion
		return nil, fmt.Errorf("reading %s: %+v", id, err)
```

File: azurerm/internal/services/apimanagement/api_management_tag_resource_test.go [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	return utils.Bool(resp.ID != nil), nil
+}
+
+func (r ApiManagementTagResource) basic(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+%s
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_tag_resource_test.go:94]:
```suggestion
%[1]s
```

File: azurerm/internal/services/apimanagement/api_management_tag_resource_test.go [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+
+resource "azurerm_api_management_tag" "test" {
+  api_management_id = azurerm_api_management.test.id
+  name              = "acctest-Op-Tag-%d"
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_tag_resource_test.go:98]:
```suggestion
  name              = "acctest-Op-Tag-%[2]d"
```

File: azurerm/internal/services/apimanagement/api_management_tag_resource_test.go [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+}
+`, ApiManagementResource{}.consumption(data), data.RandomInteger)
+}
+
+func (r ApiManagementTagResource) requiresImport(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+%s
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_tag_resource_test.go:105]:
```suggestion
%[1]s
```

File: azurerm/internal/services/apimanagement/api_management_tag_resource_test.go [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+resource "azurerm_api_management_tag" "import" {
+  api_management_id = azurerm_api_management_tag.test.api_management_id
+  name              = azurerm_api_management_tag.test.name
+}
+`, r.basic(data))
+}
+
+func (r ApiManagementTagResource) update(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+%s
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_tag_resource_test.go:116]:
```suggestion
%[1]s
```

File: azurerm/internal/services/apimanagement/api_management_tag_resource_test.go [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+
+resource "azurerm_api_management_tag" "test" {
+  api_management_id = azurerm_api_management.test.id
+  name              = "acctest-Op-Tag-%d"
```
Comment (manicminer) [inline azurerm/internal/services/apimanagement/api_management_tag_resource_test.go:120]:
```suggestion
  name              = "acctest-Op-Tag-%[2]d"
```

**Review by manicminer — CHANGES_REQUESTED** [review]
Hi @yupwei68, thanks for the PR. I've made some small suggestions below, mostly around string/error formatting. If you can fix these up and the tests pass, this should be good to merge.

File: azurerm/internal/services/apimanagement/api_management_tag_resource.go (hunk 5 above)
Comment (yupwei68) [reply azurerm/internal/services/apimanagement/api_management_tag_resource.go:89]: removed name := d.get("name").(string), so I keep displayName := d.Get("name").(string)

**Comment by yupwei68** [comment]
@manicminer Thanks for comments. Please continue reviewing.

**Review by katbyte — APPROVED** [review]
Thanks @yupwei68 - LGTM 🚀

**Pull request merged**